nxc/protocols/smb/kerberos.py

```diff
@@ -0,0 +1,282 @@
+import datetime
+import logging
+import struct
+import random
+from six import b
+
+from pyasn1.codec.der import decoder, encoder
+from pyasn1.type.univ import noValue
+
+from impacket.krb5.asn1 import AP_REQ, AS_REP, TGS_REQ, Authenticator, TGS_REP, seq_set, seq_set_iter, PA_FOR_USER_ENC, \
+    Ticket as TicketAsn1, EncTGSRepPart, PA_PAC_OPTIONS
+from impacket.krb5.types import Principal, KerberosTime, Ticket
+from impacket.krb5.kerberosv5 import sendReceive, getKerberosTGT
+from impacket.krb5.ccache import CCache
+from impacket.krb5.crypto import Key, _enctype_table, _HMACMD5
+from impacket.krb5 import constants
+
+def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash, aesKey, kdcHost, impersonate, spn, useCache, no_s4u2proxy = False):
+    logger = logging.getLogger("nxc")
+    TGT = None
+    if useCache:
+        domain, user, tgt, _ = CCache.parseFile(domain, username, f"cifs/{hostname}")
+        if TGT is None:
+            raise
+        TGT = tgt["KDC_REP"]
+        cipher = tgt["cipher"]
+        sessionKey = tgt["sessionKey"]
+    if TGT is None:
+        userName = Principal(username, type=constants.PrincipalNameType.NT_PRINCIPAL.value)
+        logger.debug("Getting TGT for user")
+        tgt, cipher, _, sessionKey = getKerberosTGT(userName, password, domain,
+                                                                lmhash, nthash,
+                                                                aesKey,
+                                                                kdcHost)
+        TGT = decoder.decode(tgt, asn1Spec=AS_REP())[0]
```

Comment:
should be snake case, but "tgt" is taken above. Maybe rename the one above and set this to "tgt"